ktor-http/ktor-http-cio/src/io/ktor/http/cio/internals/WeakTimeoutQueue.kt

```diff
@@ -142,15 +145,48 @@ class WeakTimeoutQueue(
         }
     }
 
-    @UseExperimental(InternalCoroutinesApi::class)
-    private class WeakTimeoutCoroutine<in T>(context: CoroutineContext, val delegate: Continuation<T>) :
-        AbstractCoroutine<T>(context, true), Continuation<T> {
-        override fun onCompleted(value: T) {
-            delegate.resume(value)
+    private class WeakTimeoutCoroutine<in T>(
+        context: CoroutineContext,
+        delegate: Continuation<T>,
+        val job: Job = Job(context[Job])
+    ) : Continuation<T>, Job by job, CoroutineScope {
+        override val context: CoroutineContext = context + job
+        override val coroutineContext: CoroutineContext get() = context
+
+        private val state = atomic<Continuation<T>?>(delegate)
+
+        init {
+            context[Job]?.let { parent ->
+                @UseExperimental(InternalCoroutinesApi::class)
+                parent.invokeOnCompletion(onCancelling = true) {
+                    if (it != null) {
+                        resumeWithException(it)
+                        job.cancel()
+                    }
+                }
+            }
+            job.invokeOnCompletion {
+                resumeWithException(it ?: CancellationException())
+            }
+        }
+
+        override fun resumeWith(result: Result<T>) {
+            state.getAndUpdate {
+                if (it == null) return
+                null
+            }?.let {
+                it.resumeWith(result)
+                job.cancel()
+            }
         }
 
-        override fun onCompletedExceptionally(exception: Throwable) {
-            delegate.resumeWithException(exception)
+        fun tryComplete(): Boolean {
```

Comment:
Semantics around `tryComplete` is a bit unclear, especially with`COROUTINE_SUSPENDED` part and it is very similar to our internal `tryResume`.

What you can do is to introduce your own method `tryResume + completeResume` (or indicate success of `tryResume` with nullability) and make all actions atomic instead of `tryComplete` + suspended